File: backend/app/schemas/downloads.py
"""Wire schemas for the agent download catalogue."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class AgentBuildOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    os: str
    arch: str
    version: str
    filename: str
    size_bytes: int
    sha256: str
    #: False is rendered, never hidden. An unsigned binary means Gatekeeper or
    #: SmartScreen will interrupt the install, and the page says so before the
    #: click rather than leaving the user to meet it alone.
    signed: bool
    signing: str
    built_at: datetime | None
    download_url: str


class DownloadTicketOut(BaseModel):
    """How long the download credential this request just minted will last.

    The credential itself is **not** here: it goes back as an HttpOnly cookie
    (`app/security/cookies.py`), because a query string is the one place a URL
    reliably leaks — access logs, browser history, `Referer`. The page needs
    only the lifetime, so it can re-mint before a backgrounded tab's ticket
    lapses.
    """

    expires_in: int


class AgentDownloadsOut(BaseModel):
    """The catalogue, including the case where it is empty.

    `unavailable_reason` exists so "nothing published yet" is a state the UI can
    render honestly, rather than an empty list it has to guess about.
    """

    configured: bool
    generated_at: datetime | None
    builds: list[AgentBuildOut]
    unavailable_reason: str | None
    #: Echoed back so the page can name the exact command that builds the
    #: version this server expects.
    source_build_command: str = "cd agent && pip install -e '.[build]' && python build/build.py"

File: web/src/types/downloads.ts
/** Mirrors app/schemas/downloads.py. */

export type BuildOs = "macos" | "linux" | "windows" | "android"

export interface AgentBuild {
  os: BuildOs
  arch: "x64" | "arm64" | "universal"
  version: string
  filename: string
  size_bytes: number
  sha256: string
  /** False means Gatekeeper or SmartScreen will interrupt the install. */
  signed: boolean
  signing: string
  built_at: string | null
  /** Absolute when a release host is configured; otherwise a path on this API. */
  download_url: string
}

export interface AgentDownloads {
  configured: boolean
  generated_at: string | null
  builds: AgentBuild[]
  /** Non-null whenever `builds` is empty — why, in words the page can render. */
  unavailable_reason: string | null
  source_build_command: string
}
